File: mcp_server_erlang/src/mcp_server.erl
%%%-------------------------------------------------------------------
%%% @doc MCP Server - Main server implementation
%%%
%%% Model Context Protocol server for Erlang/OTP.
%%% Communicates via stdio using JSON-RPC 2.0.
%%%
%%% Copyright (C) 2024 ALGT Project
%%%
%%% Usage:
%%%   erl -pa _build/default/lib/*/ebin -s mcp_server start
%%%-------------------------------------------------------------------
-module(mcp_server).

-behaviour(gen_server).

%% API
-export([
    start/0,
    start_link/0,
    stop/0
]).

%% gen_server callbacks
-export([
    init/1,
    handle_call/3,
    handle_cast/2,
    handle_info/2,
    terminate/2,
    code_change/3
]).

-define(SERVER, ?MODULE).
-define(PROTOCOL_VERSION, <<"2024-11-05">>).
-define(SERVER_NAME, <<"algt-mcp-server-erlang">>).
-define(SERVER_VERSION, <<"0.1.0">>).

-record(state, {
    initialized = false :: boolean()
}).

%%%===================================================================
%%% API
%%%===================================================================

%%%-------------------------------------------------------------------
%%% @doc Start the MCP server (called from command line)
%%%-------------------------------------------------------------------
start() ->
    %% Start required applications
    application:ensure_all_started(jsx),

    %% Start the tools registry
    {ok, _} = mcp_tools:start_link(),

    %% Register ALGT tools
    mcp_algt_tools:register_all(),

    %% Start the server and enter message loop
    {ok, Pid} = start_link(),
    log_info("ALGT MCP Server (Erlang) starting..."),

    %% Enter the message loop
    message_loop(Pid),

    %% Halt when done
    init:stop().

%%%-------------------------------------------------------------------
%%% @doc Start the server as a gen_server
%%%-------------------------------------------------------------------
start_link() ->
    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).

%%%-------------------------------------------------------------------
%%% @doc Stop the server
%%%-------------------------------------------------------------------
stop() ->
    gen_server:stop(?SERVER).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

init([]) ->
    {ok, #state{}}.

handle_call({handle_message, Message}, _From, State) ->
    {Response, NewState} = process_message(Message, State),
    {reply, Response, NewState};

handle_call(_Request, _From, State) ->
    {reply, {error, unknown_request}, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%%===================================================================
%%% Message Loop
%%%===================================================================

%%%-------------------------------------------------------------------
%%% @doc Main message processing loop
%%%-------------------------------------------------------------------
message_loop(Pid) ->
    case mcp_protocol:read_message() of
        eof ->
            log_info("End of input, shutting down"),
            ok;
        {error, Reason} ->
            log_error("Read error: ~p", [Reason]),
            ok;
        {ok, Message} ->
            case gen_server:call(Pid, {handle_message, Message}) of
                {response, Response} ->
                    mcp_protocol:write_response(Response);
                no_response ->
                    ok;
                shutdown ->
                    ok
            end,
            message_loop(Pid)
    end.

%%%===================================================================
%%% Message Processing
%%%===================================================================

%%%-------------------------------------------------------------------
%%% @doc Process a JSON-RPC message
%%%-------------------------------------------------------------------
-spec process_message(map(), #state{}) ->
    {{response, map()} | no_response | shutdown, #state{}}.
process_message(Message, State) ->
    Id = maps:get(<<"id">>, Message, null),
    Method = maps:get(<<"method">>, Message, <<>>),
    Params = maps:get(<<"params">>, Message, #{}),

    log_info("Received: ~s (id=~p)", [Method, Id]),

    case Method of
        <<"initialize">> ->
            handle_initialize(Id, Params, State);
        <<"initialized">> ->
            handle_initialized(State);
        <<"shutdown">> ->
            handle_shutdown(Id, State);
        <<"tools/list">> ->
            handle_tools_list(Id, State);
        <<"tools/call">> ->
            handle_tools_call(Id, Params, State);
        <<"ping">> ->
            handle_ping(Id, State);
        _ when Id =/= null ->
            %% Unknown method with id -> error response
            ErrorMsg = iolist_to_binary(
                io_lib:format("Unknown method: ~s", [Method])),
            Response = mcp_protocol:error_response(
                Id,
                mcp_protocol:method_not_found(),
                ErrorMsg
            ),
            {{response, Response}, State};
        _ ->
            %% Unknown notification -> ignore
            log_info("Ignoring unknown notification: ~s", [Method]),
            {no_response, State}
    end.

%%%-------------------------------------------------------------------
%%% @doc Handle initialize request
%%%-------------------------------------------------------------------
handle_initialize(Id, Params, State) ->
    %% Log client info if present
    case maps:get(<<"clientInfo">>, Params, undefined) of
        undefined -> ok;
        ClientInfo -> log_info("Client: ~p", [ClientInfo])
    end,

    Response = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"result">> => #{
            <<"protocolVersion">> => ?PROTOCOL_VERSION,
            <<"capabilities">> => #{
                <<"tools">> => #{
                    <<"listChanged">> => false
                }
            },
            <<"serverInfo">> => #{
                <<"name">> => ?SERVER_NAME,
                <<"version">> => ?SERVER_VERSION
            }
        }
    },
    {{response, Response}, State#state{initialized = true}}.

%%%-------------------------------------------------------------------
%%% @doc Handle initialized notification
%%%-------------------------------------------------------------------
handle_initialized(State) ->
    log_info("Client initialized successfully"),
    {no_response, State}.

%%%-------------------------------------------------------------------
%%% @doc Handle shutdown request
%%%-------------------------------------------------------------------
handle_shutdown(Id, State) ->
    log_info("Shutdown requested"),
    Response = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"result">> => null
    },
    {{response, Response}, State}.

%%%-------------------------------------------------------------------
%%% @doc Handle ping request
%%%-------------------------------------------------------------------
handle_ping(Id, State) ->
    Response = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"result">> => #{}
    },
    {{response, Response}, State}.

%%%-------------------------------------------------------------------
%%% @doc Handle tools/list request
%%%-------------------------------------------------------------------
handle_tools_list(Id, State) ->
    Tools = mcp_tools:list_tools(),
    Response = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"result">> => #{
            <<"tools">> => Tools
        }
    },
    {{response, Response}, State}.

%%%-------------------------------------------------------------------
%%% @doc Handle tools/call request
%%%-------------------------------------------------------------------
handle_tools_call(Id, Params, State) ->
    Response = case {maps:get(<<"name">>, Params, undefined),
                     maps:get(<<"arguments">>, Params, undefined)} of
        {undefined, _} ->
            mcp_protocol:error_response(
                Id,
                mcp_protocol:invalid_params(),
                <<"Missing required parameter: name">>
            );
        {_, undefined} ->
            mcp_protocol:error_response(
                Id,
                mcp_protocol:invalid_params(),
                <<"Missing required parameter: arguments">>
            );
        {ToolName, Arguments} ->
            log_info("Calling tool: ~s", [ToolName]),
            ToolResult = mcp_tools:call_tool(ToolName, Arguments),
            #{
                <<"jsonrpc">> => <<"2.0">>,
                <<"id">> => Id,
                <<"result">> => ToolResult
            }
    end,
    {{response, Response}, State}.

%%%===================================================================
%%% Logging
%%%===================================================================

log_info(Format, Args) ->
    Timestamp = calendar:system_time_to_rfc3339(
        erlang:system_time(second)),
    io:format(standard_error, "[~s] INFO: " ++ Format ++ "~n",
              [Timestamp | Args]).

log_info(Message) ->
    log_info("~s", [Message]).

log_error(Format, Args) ->
    Timestamp = calendar:system_time_to_rfc3339(
        erlang:system_time(second)),
    io:format(standard_error, "[~s] ERROR: " ++ Format ++ "~n",
              [Timestamp | Args]).

File: mcp_servers/elixir/lib/mcp_server.ex
defmodule McpServer do
  @moduledoc """
  Main MCP (Model Context Protocol) server implementation for Elixir.

  Communicates via stdio using JSON-RPC 2.0.

  ## Usage

      # Start from command line:
      mix run --no-halt -e "McpServer.start()"

      # Or use the escript:
      ./mcp_server
  """

  use GenServer

  alias McpServer.Protocol
  alias McpServer.Tools
  alias McpServer.AlgtTools

  @protocol_version "2024-11-05"
  @server_name "algt-mcp-server-elixir"
  @server_version "0.1.0"

  defstruct initialized: false

  # Public API

  @doc """
  Start the MCP server (entry point for command line).
  """
  def start do
    # Start the tools registry
    {:ok, _} = Tools.start_link()

    # Register ALGT tools
    AlgtTools.register_all()

    # Start the server
    {:ok, pid} = start_link()

    log_info("ALGT MCP Server (Elixir) starting...")

    # Enter the message loop
    message_loop(pid)
  end

  @doc """
  Start the server as a GenServer.
  """
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  # GenServer Callbacks

  @impl true
  def init(_opts) do
    {:ok, %__MODULE__{}}
  end

  @impl true
  def handle_call({:handle_message, message}, _from, state) do
    {response, new_state} = process_message(message, state)
    {:reply, response, new_state}
  end

  # Message Loop

  defp message_loop(pid) do
    case Protocol.read_message() do
      :eof ->
        log_info("End of input, shutting down")
        :ok

      {:error, reason} ->
        log_error("Read error: #{inspect(reason)}")
        :ok

      {:ok, message} ->
        case GenServer.call(pid, {:handle_message, message}) do
          {:response, response} ->
            Protocol.write_response(response)

          :no_response ->
            :ok

          :shutdown ->
            :ok
        end

        message_loop(pid)
    end
  end

  # Message Processing

  defp process_message(message, state) do
    id = Map.get(message, "id")
    method = Map.get(message, "method", "")
    params = Map.get(message, "params", %{})

    log_info("Received: #{method} (id=#{inspect(id)})")

    case method do
      "initialize" ->
        handle_initialize(id, params, state)

      "initialized" ->
        handle_initialized(state)

      "shutdown" ->
        handle_shutdown(id, state)

      "tools/list" ->
        handle_tools_list(id, state)

      "tools/call" ->
        handle_tools_call(id, params, state)

      "ping" ->
        handle_ping(id, state)

      _ when not is_nil(id) ->
        # Unknown method with id -> error response
        response =
          Protocol.error_response(
            id,
            Protocol.method_not_found(),
            "Unknown method: #{method}"
          )

        {{:response, response}, state}

      _ ->
        # Unknown notification -> ignore
        log_info("Ignoring unknown notification: #{method}")
        {:no_response, state}
    end
  end

  # Handler Functions

  defp handle_initialize(id, params, state) do
    # Log client info if present
    case Map.get(params, "clientInfo") do
      nil -> :ok
      client_info -> log_info("Client: #{inspect(client_info)}")
    end

    response = %{
      "jsonrpc" => "2.0",
      "id" => id,
      "result" => %{
        "protocolVersion" => @protocol_version,
        "capabilities" => %{
          "tools" => %{
            "listChanged" => false
          }
        },
        "serverInfo" => %{
          "name" => @server_name,
          "version" => @server_version
        }
      }
    }

    {{:response, response}, %{state | initialized: true}}
  end

  defp handle_initialized(state) do
    log_info("Client initialized successfully")
    {:no_response, state}
  end

  defp handle_shutdown(id, state) do
    log_info("Shutdown requested")

    response = %{
      "jsonrpc" => "2.0",
      "id" => id,
      "result" => nil
    }

    {{:response, response}, state}
  end

  defp handle_ping(id, state) do
    response = %{
      "jsonrpc" => "2.0",
      "id" => id,
      "result" => %{}
    }

    {{:response, response}, state}
  end

  defp handle_tools_list(id, state) do
    tools = Tools.list()

    response = %{
      "jsonrpc" => "2.0",
      "id" => id,
      "result" => %{
        "tools" => tools
      }
    }

    {{:response, response}, state}
  end

  defp handle_tools_call(id, params, state) do
    response =
      case {Map.get(params, "name"), Map.get(params, "arguments")} do
        {nil, _} ->
          Protocol.error_response(
            id,
            Protocol.invalid_params(),
            "Missing required parameter: name"
          )

        {_, nil} ->
          Protocol.error_response(
            id,
            Protocol.invalid_params(),
            "Missing required parameter: arguments"
          )

        {tool_name, arguments} ->
          log_info("Calling tool: #{tool_name}")
          tool_result = Tools.call(tool_name, arguments)

          %{
            "jsonrpc" => "2.0",
            "id" => id,
            "result" => tool_result
          }
      end

    {{:response, response}, state}
  end

  # Logging

  defp log_info(message) do
    timestamp = DateTime.utc_now() |> DateTime.to_iso8601()
    IO.write(:stderr, "[#{timestamp}] INFO: #{message}\n")
  end

  defp log_error(message) do
    timestamp = DateTime.utc_now() |> DateTime.to_iso8601()
    IO.write(:stderr, "[#{timestamp}] ERROR: #{message}\n")
  end
end
